hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/SqmSetReturningFunction.java

```diff
@@ -83,15 +84,30 @@ public void appendHqlString(StringBuilder hql, SqmRenderContext context) {
 		hql.append( functionName );
 		if ( arguments.isEmpty() ) {
 			hql.append( "()" );
-			return;
 		}
-		hql.append( '(' );
-		arguments.get( 0 ).appendHqlString( hql, context );
-		for ( int i = 1; i < arguments.size(); i++ ) {
-			hql.append( ", " );
-			arguments.get( i ).appendHqlString( hql, context );
+		else {
+			hql.append( '(' );
+			arguments.get( 0 ).appendHqlString( hql, context );
+			for ( int i = 1; i < arguments.size(); i++ ) {
+				hql.append( ", " );
+				arguments.get( i ).appendHqlString( hql, context );
+			}
+			hql.append( ')' );
 		}
+	}
 
-		hql.append( ')' );
+	@Override
+	// TODO: override on all subtypes
+	public boolean equals(Object other) {
+		return other instanceof SqmSetReturningFunction<?> that
+			&& Objects.equals( this.functionName, that.functionName )
+			&& Objects.equals( this.arguments, that.arguments )
+			&& this.getClass() == that.getClass()
+			&& Objects.equals( this.toHqlString(), that.toHqlString() );
```

Comment:
Again, to avoid aliasing conflicts and because HQL is not necessary AFAIU, it might be best to simply check all the constituent parts are equal.

Reply:
Right, but I belike that means overriding this on every subtype, as indicated in the comment.

Reply:
True, that will require that every function subtype implements this properly.

Reply:
We will need to get around to doing that. But this is OK for now, it seems to me.